File: packages/onyx-sdk-python/onyx_escrow/api_keys.py
"""API key management for Onyx SDK.

Provides CRUD operations for B2B API keys.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .client import OnyxClient


class ApiKeyManager:
    """Manages API key lifecycle.

    Accessed via ``client.api_keys``.

    Example::

        async with OnyxClient(api_key="nxs_...") as client:
            keys = await client.api_keys.list()
            print(f"Active keys: {keys['total']}")
    """

    def __init__(self, client: OnyxClient) -> None:
        self._client = client

    async def create(
        self,
        name: str,
        *,
        expires_at: str | None = None,
        metadata: str | None = None,
        csrf_token: str = "",
    ) -> dict[str, Any]:
        """Create a new API key.

        The raw key is only returned once at creation time.

        Args:
            name: Human-readable name for the key.
            expires_at: Optional expiration (ISO 8601: YYYY-MM-DD HH:MM:SS).
            metadata: Optional JSON metadata string.
            csrf_token: CSRF token (required for session-authenticated requests).

        Returns:
            Creation response including the raw API key.
        """
        payload: dict[str, Any] = {"name": name, "csrf_token": csrf_token}
        if expires_at is not None:
            payload["expires_at"] = expires_at
        if metadata is not None:
            payload["metadata"] = metadata

        return await self._client._request(
            "POST", "/api/api-keys", json=payload
        )

    async def list(self) -> dict[str, Any]:
        """List all API keys for the authenticated user.

        Returns:
            Dict with ``keys`` list and ``total`` count.
        """
        return await self._client._request("GET", "/api/api-keys")

    async def get(self, key_id: str) -> dict[str, Any]:
        """Get details of a specific API key.

        Args:
            key_id: The API key identifier.

        Returns:
            API key info.
        """
        return await self._client._request("GET", f"/api/api-keys/{key_id}")

    async def revoke(self, key_id: str) -> dict[str, Any]:
        """Revoke (deactivate) an API key.

        The key will no longer be usable for authentication.

        Args:
            key_id: The API key identifier.

        Returns:
            Revocation confirmation.
        """
        return await self._client._request("DELETE", f"/api/api-keys/{key_id}")

    async def delete(self, key_id: str) -> dict[str, Any]:
        """Permanently delete an API key.

        Args:
            key_id: The API key identifier.

        Returns:
            Deletion confirmation.
        """
        return await self._client._request(
            "DELETE", f"/api/api-keys/{key_id}/permanent"
        )

File: packages/nexus-sdk-js/src/apikeys.ts
/**
 * NEXUS SDK API Key Management
 *
 * Session-authenticated endpoints for managing B2B API keys.
 */

import type { ApiKey, CreateApiKeyParams, ApiKeyCreationResponse } from './types.js';

type RequestFn = <T>(method: string, path: string, body?: unknown) => Promise<T>;

/**
 * API Key management resource
 *
 * @example
 * ```typescript
 * // Create a new API key
 * const result = await client.apiKeys.create({
 *   name: 'Production Key',
 *   csrf_token: csrfToken,
 * });
 * console.log(`Key: ${result.raw_key}`); // Only shown once
 *
 * // List all keys
 * const { keys } = await client.apiKeys.list();
 * ```
 */
export class ApiKeyResource {
  private readonly request: RequestFn;

  constructor(request: RequestFn) {
    this.request = request;
  }

  /**
   * Create a new API key
   *
   * POST /api/api-keys
   *
   * The raw key is only returned once at creation time.
   */
  async create(params: CreateApiKeyParams): Promise<{
    message: string;
    key: ApiKeyCreationResponse;
  }> {
    return this.request('POST', '/api-keys', params);
  }

  /**
   * List all API keys for the authenticated user
   *
   * GET /api/api-keys
   */
  async list(): Promise<{ keys: ApiKey[]; total: number }> {
    return this.request('GET', '/api-keys');
  }

  /**
   * Get details of a specific API key
   *
   * GET /api/api-keys/{id}
   */
  async get(keyId: string): Promise<ApiKey> {
    return this.request<ApiKey>('GET', `/api/api-keys/${keyId}`);
  }

  /**
   * Revoke (deactivate) an API key
   *
   * DELETE /api/api-keys/{id}
   */
  async revoke(keyId: string): Promise<{ message: string; key_id: string }> {
    return this.request('DELETE', `/api/api-keys/${keyId}`);
  }

  /**
   * Permanently delete an API key
   *
   * DELETE /api/api-keys/{id}/permanent
   */
  async delete(keyId: string): Promise<{ message: string; key_id: string }> {
    return this.request('DELETE', `/api/api-keys/${keyId}/permanent`);
  }
}
